syft/pkg/cataloger/golang/parse_go_bin.go

```diff
@@ -39,57 +63,112 @@ func newGoBinaryPackage(name, version, h1Digest, goVersion, architecture string,
 	return p
 }
 
-func parseGoBin(location source.Location, reader io.ReadCloser, opener exeOpener) (pkgs []pkg.Package, err error) {
-	var exes []exe
-	// it has been found that there are stdlib paths within openExe that can panic. We want to prevent this behavior
-	// bubbling up and halting execution. For this reason we try to recover from any panic and return an error.
-	defer func() {
-		if r := recover(); r != nil {
-			err = fmt.Errorf("recovered from panic while parse go binary at %q: %+v", location.RealPath, r)
+func setArch(readers []io.ReaderAt, builds []*debug.BuildInfo) {
+	if len(readers) != len(builds) {
+		log.Errorf("golang cataloger: bin parsing: number of builds and readers doesn't match")
+		return
+	}
+
+	for _, build := range builds {
+		if getGOARCH(build.Settings) != "" {
+			return
+		}
+		break
+	}
+
+	for i, r := range readers {
+		a, err := getGOARCHFromBin(r)
+		if err != nil || a == "" {
+			log.Warnf("golang cataloger: bin parsing: getting arch from binary: %v", err)
+			continue
 		}
-	}()
 
-	// Identify if bin was compiled by go
-	exes, err = opener(reader)
-	if err != nil {
-		return pkgs, err
+		builds[i].Settings = append(builds[i].Settings, debug.BuildSetting{Key: GOARCH, Value: a})
```

Comment:
we shouldn't be modifying the build settings; the architecture should be raised up separately